src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -1235,15 +1294,22 @@ class SegmentCleaner : public SegmentProvider {
     } else {
       assert(segments[segment].is_empty());
       assert(segments.get_empty_segments(segment.device_id()) > 0);
-      segments.segment_used(segment, true);
-      stats.avail_bytes -= segments[segment.device_id()]->segment_size;
+      assert(stats.empty_segments > 0);
       stats.empty_segments--;
     }
-    crimson::get_logger(ceph_subsys_seastore).debug(
-      "mark_closed: device: {} empty_segments: {}",
-      segment.device_id(),
-      segments.get_empty_segments(segment.device_id()));
-    segments[segment].set_closed();
+    segments.segment_closed(segment);
+    stats.avail_bytes = segments.get_available_bytes();
+    stats.opened_segments = segments.get_opened_segments();
```

Comment:
Can we remove both `stats.avail_bytes` and `stats.opened_segments` and always replace them by calling to `segment_info_set_t` interfaces directly?

Reply:
The stats are for the seastar.metrics

Reply:
Metrics don't require variables, they can also be registerred through lambdas:
https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/cache.cc#L468-L474

If we use stats.avail_bytes and stats.opened_segments, we still need them to be updated each time to get the most accurate values, right?

Reply:
I think there are two considerations for a metric (whether it is a variable or through a lambda):
* It should be lightweight to be maintained/updated during I/O;
* It should be lightweight to be fetched because we may need to get a lot of metrics for each collection and we also don't want this to affect the test workload;

This needs careful implementations, so I'm fine to leave it to a follow-up PR.

Reply:
> 

I think I can do this in this pr